src/App.php

```diff
@@ -763,12 +830,9 @@ public function requireCSS($url)
      * @throws \atk4\core\Exception
      * @throws ExitApplicationException
      */
-    public function redirect($page)
+    public function redirect($page, bool $permanent = false): void
     {
-        header('Location: ' . $this->url($page));
-
-        $this->run_called = true; // prevent shutdown function from triggering.
```

Comment:
i remember this `run_called` property was used somewhere in legacy project but probably it's OK to not use it there.

Reply:
Yes, and the new terminateHTML calls terminate which sets it as previously, only at one place. The same again, the more you duplicate, the more you have to worry about :D